Continuous-integration tests keep failing, blocking any new pull request

I stumbled over this issue when trying to submit PR #278. This is essentially a copy of what I've reported there:
* The tests fail at the second version of python that is going to be tested, no matter which version it happens to be.
* Therefore, if one modifies `ci.yml` to test only a single version of python, the tests will pass.
* The failure occurs at the `Run pytest` stage, the error message (at lines 31-33) is the following
   - 31 ` /home/runner/work/_temp/6ab19913-178e-4b8f-92bf-dae56f3545bf.sh: line 1:  1858 Segmentation fault      (core dumped) PPAFM_RECOMPILE=1 pytest tests examples -v --cov --cov-report json`
   - 32 `examples/PTCDA_single/example_ptcda.py::example_ptcda_single`
   - 33 `Error: Process completed with exit code 139.`
 * The tests fail even for a branch that is an exact copy of the current `main` branch, so it is not an effect of any changes I tried to implement.
 
I don't have much of an idea how to debug and fix these issues with the GitHub 'workflow' functionality, so I would appreciate any help from you guys who know more about it.

__Reposted from__ #295 :

It looks to me, like a bug in GitHub or lowering down the policy, since all tests have died at 4 minutes marks. I did not changed any part of the "active code", only README and a commented part. Anyway the test stopped with the following error for python3.7:
```
Run PPAFM_RECOMPILE=1 pytest tests examples/PTCDA_Hartree_dz2 -v --cov --cov-report json
============================= test session starts ==============================
platform linux -- Python 3.7.17, pytest-7.4.4, pluggy-1.2.0 -- /opt/hostedtoolcache/Python/3.7.17/x64/bin/python
cachedir: .pytest_cache
rootdir: /home/runner/work/ppafm/ppafm
configfile: pyproject.toml
plugins: cov-4.1.0
collecting ... collected 12 items

tests/test_afmulator.py::test_afmulator_save_load PASSED                 [  8%]
tests/test_atomicUtils.py::test_ZsToElems PASSED                         [ 16%]
tests/test_common.py::test_get_df_weight PASSED                          [ 25%]
tests/test_common.py::test_get_simple_df_weight PASSED                   [ 33%]
tests/test_common.py::test_sphere_tangent_space PASSED                   [ 41%]
tests/test_datagrid.py::test_power PASSED                                [ 50%]
tests/test_datagrid.py::test_tip_interp PASSED                           [ 58%]
tests/test_generator.py::test_GeneratorAFMtrainer PASSED                 [ 66%]
tests/test_io.py::test_xyz PASSED                                        [ 75%]
tests/test_io.py::test_parse_comment_ase PASSED                          [ 83%]
tests/test_io.py::test_load_aims PASSED                                  [ 91%]
Error: The operation was canceled.
```

The python 3.11 seems an interesting one:
```
Run PPAFM_RECOMPILE=1 pytest tests examples/PTCDA_Hartree_dz2 -v --cov --cov-report json
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.3.2, pluggy-1.5.0 -- /opt/hostedtoolcache/Python/3.11.9/x64/bin/python
cachedir: .pytest_cache
rootdir: /home/runner/work/ppafm/ppafm
configfile: pyproject.toml
plugins: cov-5.0.0
collecting ... collected 12 items

tests/test_afmulator.py::test_afmulator_save_load PASSED                 [  8%]
tests/test_atomicUtils.py::test_ZsToElems PASSED                         [ 16%]
tests/test_common.py::test_get_df_weight PASSED                          [ 25%]
tests/test_common.py::test_get_simple_df_weight PASSED                   [ 33%]
tests/test_common.py::test_sphere_tangent_space PASSED                   [ 41%]
tests/test_datagrid.py::test_power PASSED                                [ 50%]
tests/test_datagrid.py::test_tip_interp PASSED                           [ 58%]
tests/test_generator.py::test_GeneratorAFMtrainer PASSED                 [ 66%]
tests/test_io.py::test_xyz PASSED                                        [ 75%]
tests/test_io.py::test_parse_comment_ase PASSED                          [ 83%]
tests/test_io.py::test_load_aims PASSED                                  [ 91%]
/home/runner/work/_temp/a2a60b1b-2468-44ab-93c7-89e20eb061a1.sh: line 1:  1839 Segmentation fault      (core dumped) PPAFM_RECOMPILE=1 pytest tests examples/PTCDA_Hartree_dz2 -v --cov --cov-report json
examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py::example_ptcda_hartree 
Error: Process completed with exit code 139.
```
Similarly 3.12 is having problems with the PTCDA_Hartree_dz2 example:
```
Run PPAFM_RECOMPILE=1 pytest tests examples/PTCDA_Hartree_dz2 -v --cov --cov-report json
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.3.2, pluggy-1.5.0 -- /opt/hostedtoolcache/Python/3.12.4/x64/bin/python
cachedir: .pytest_cache
rootdir: /home/runner/work/ppafm/ppafm
configfile: pyproject.toml
plugins: cov-5.0.0
collecting ... collected 12 items

tests/test_afmulator.py::test_afmulator_save_load PASSED                 [  8%]
tests/test_atomicUtils.py::test_ZsToElems PASSED                         [ 16%]
tests/test_common.py::test_get_df_weight PASSED                          [ 25%]
tests/test_common.py::test_get_simple_df_weight PASSED                   [ 33%]
tests/test_common.py::test_sphere_tangent_space PASSED                   [ 41%]
tests/test_datagrid.py::test_power PASSED                                [ 50%]
tests/test_datagrid.py::test_tip_interp PASSED                           [ 58%]
tests/test_generator.py::test_GeneratorAFMtrainer PASSED                 [ 66%]
tests/test_io.py::test_xyz PASSED                                        [ 75%]
tests/test_io.py::test_parse_comment_ase PASSED                          [ 83%]
tests/test_io.py::test_load_aims PASSED                                  [ 91%]
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_algorithms_reduce_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_algorithms_reduce_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_algorithms_scan_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_algorithms_scan_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_algorithms_transpose_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_algorithms_transpose_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_cluda_functions_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_cluda_functions_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_cluda_kernel_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_cluda_kernel_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_cluda_vsize_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_cluda_vsize_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_fft_fft_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_fft_fft_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/coverage/report_core.py:115: CoverageWarning: Couldn't parse '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_fft_fftshift_mako': No source for code: '/home/runner/work/ppafm/ppafm/_opt_hostedtoolcache_Python_3_12_4_x64_lib_python3_12_site_packages_reikna_fft_fftshift_mako'. (couldnt-parse)
  coverage._warn(msg, slug="couldnt-parse")
examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py::example_ptcda_hartree PASSED [100%]

=============================== warnings summary ===============================
../../../../../opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/pytools/persistent_dict.py:59
  /opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/pytools/persistent_dict.py:59: UserWarning: Unable to import recommended hash 'siphash24.siphash13', falling back to 'hashlib.sha256'. Run 'python3 -m pip install siphash24' to install the recommended hash.
    warn("Unable to import recommended hash 'siphash24.siphash13', "

ppafm/cli/plot_results.py:12
  /home/runner/work/ppafm/ppafm/ppafm/cli/plot_results.py:12: MatplotlibDeprecationWarning: Auto-close()ing of figures upon backend switching is deprecated since 3.8 and will be removed in 3.10.  To suppress this warning, explicitly call plt.close('all') first.
    mpl.use("Agg")

tests/test_afmulator.py::test_afmulator_save_load
tests/test_afmulator.py::test_afmulator_save_load
tests/test_afmulator.py::test_afmulator_save_load
examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py::example_ptcda_hartree
examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py::example_ptcda_hartree
  /home/runner/work/ppafm/ppafm/ppafm/fieldFFT.py:232: PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray.
    return np.matrix(Lmat)

examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py::example_ptcda_hartree
  /home/runner/work/ppafm/ppafm/ppafm/fieldFFT.py:18: PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray.
    return np.matrix(lvec[1:])

examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py::example_ptcda_hartree
  /opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/numpy/matrixlib/defmatrix.py:70: PendingDeprecationWarning: the matrix subclass is not the recommended way to represent matrices or deal with linear algebra (see https://docs.scipy.org/doc/numpy/user/numpy-for-matlab-users.html). Please adjust your code to use regular ndarray.
    return matrix(data, dtype=dtype, copy=False)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.12.4-final-0 -----------
Coverage JSON written to file coverage.json

================== 12 passed, 9 warnings in 203.48s (0:03:23) ==================
FileRead program: reading Q-0.10K0.50/OutFz.xsf file
XYZ dimensions are 62 202 202
Reading DONE
Error: The operation was canceled.
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -7,7 +7,7 @@ jobs:
         runs-on: ubuntu-latest
         strategy:
             matrix:
-                python-version: ['3.7', '3.9', '3.10', '3.11', '3.12']
+                python-version: ['3.7', '3.8', '3.9', '3.10', '3.11', '3.12']
 
         steps:
             - name: Check out repository
@@ -24,8 +24,7 @@ jobs:
                   pip install pocl-binary-distribution
 
             - name: Run pytest
-              # run: PPAFM_RECOMPILE=1 pytest tests examples -v --cov --cov-report json # Re-enable this when #232 is resolved
-              run: PPAFM_RECOMPILE=1 pytest tests examples/PTCDA_Hartree_dz2 -v --cov --cov-report json
+              run: PPAFM_RECOMPILE=1 pytest tests examples -v --cov --cov-report json
 
             - name: Upload coverage report
               uses: codecov/codecov-action@v3

--- a/examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py
+++ b/examples/PTCDA_Hartree_dz2/example_ptcda_hartree.py
@@ -14,8 +14,9 @@ def example_ptcda_hartree():
     # Change directory to the location of this script
     os.chdir(script_location)
 
-    subprocess.run(["wget", "--no-check-certificate", "https://www.dropbox.com/s/18eg89l89npll8x/LOCPOT.xsf.zip"])
-    subprocess.run(["unzip", "LOCPOT.xsf.zip"])
+    if not Path("LOCPOT.xsf").exists():
+        subprocess.run(["wget", "--no-check-certificate", "https://www.dropbox.com/s/18eg89l89npll8x/LOCPOT.xsf.zip"])
+        subprocess.run(["unzip", "LOCPOT.xsf.zip"])
 
     generate_ljff(["--input", "LOCPOT.xsf"])
     generate_elff(["--input", "LOCPOT.xsf", "--tip", "dz2"])

--- a/ppafm/HighLevel.py
+++ b/ppafm/HighLevel.py
@@ -103,7 +103,7 @@ def relaxedScan3D(xTips, yTips, zTips, trj=None, bF3d=False):
     return fzs, PPpos
 
 
-def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False):
+def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False, spline_parameters=None):
     if verbose > 0:
         print(">>BEGIN: relaxedScan3D_omp()")
     if verbose > 0:
@@ -117,7 +117,7 @@ def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False):
     rTips[:, :, :, 0] = xTips[:, None, None]
     rTips[:, :, :, 1] = yTips[None, :, None]
     rTips[:, :, :, 2] = zTips[::-1][None, None, :]
-    core.relaxTipStrokes_omp(rTips, rs, fs)
+    core.relaxTipStrokes_omp(rTips, rs, fs, spline_parameters=spline_parameters)
     # rs[:,:,:,:] = rs[:,:,::-1,:].transpose(2,1,0,3).copy()
     # fs[:,:,:,:] = fs[:,:,::-1,:]
     # fzs = fs[:,:,::-1,2].transpose(2,1,0).copy()
@@ -131,26 +131,22 @@ def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False):
     return fzs, rs
 
 
-def perform_relaxation(lvec, FFLJ, FFel=None, FFpauli=None, FFboltz=None, FFkpfm_t0sV=None, FFkpfm_tVs0=None, tipspline=None, bPPdisp=False, bFFtotDebug=False, parameters=None):
+def perform_relaxation(
+    lvec,
+    FFLJ,
+    FFel=None,
+    FFpauli=None,
+    FFboltz=None,
+    FFkpfm_t0sV=None,
+    FFkpfm_tVs0=None,
+    spline_parameters=None,
+    bPPdisp=False,
+    bFFtotDebug=False,
+    parameters=None,
+):
     global FF  # We need FF global otherwise it is garbage collected and program crashes inside C++ e.g. in stiffnessMatrix()
     if verbose > 0:
         print(">>>BEGIN: perform_relaxation()")
-    if tipspline is not None:
-        try:
-            if verbose > 0:
-                print(" loading tip spline from " + tipspline)
-            S = np.genfromtxt(tipspline)
-            xs = S[:, 0].copy()
-            if verbose > 0:
-                print("xs: ", xs)
-            ydys = S[:, 1:].copy()
-            if verbose > 0:
-                print("ydys: ", ydys)
-            core.setTipSpline(xs, ydys)
-        except:
-            if verbose > 0:
-                print("cannot load tip spline from " + tipspline)
-            sys.exit()
     xTips, yTips, zTips, lvecScan = PPU.prepareScanGrids(parameters=parameters)
     FF = FFLJ.copy()
     if FFel is not None:
@@ -180,7 +176,7 @@ def perform_relaxation(lvec, FFLJ, FFel=None, FFpauli=None, FFboltz=None, FFkpfm
     if parameters.tiltedScan:
         trj = trjByDir(len(zTips), d=parameters.scanTilt, p0=[0.0, 0.0, parameters.scanMin[2] - lvec[0, 2]])
     # fzs, PPpos = relaxedScan3D(xTips - lvec[0, 0], yTips - lvec[0, 1], zTips - lvec[0, 2], trj=trj, bF3d=parameters.tiltedScan)
-    fzs, PPpos = relaxedScan3D_omp(xTips - lvec[0, 0], yTips - lvec[0, 1], zTips - lvec[0, 2], trj=trj, bF3d=parameters.tiltedScan)
+    fzs, PPpos = relaxedScan3D_omp(xTips - lvec[0, 0], yTips - lvec[0, 1], zTips - lvec[0, 2], trj=trj, bF3d=parameters.tiltedScan, spline_parameters=spline_parameters)
 
     # transform probe-particle positions back to the original coordinates
     PPpos[:, :, :, 0] += lvec[0, 0]

--- a/ppafm/HighLevel.py
+++ b/ppafm/HighLevel.py
@@ -103,7 +103,7 @@ def relaxedScan3D(xTips, yTips, zTips, trj=None, bF3d=False):
     return fzs, PPpos
 
 
-def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False, spline_parameters=None):
+def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False, tip_spline=None):
     if verbose > 0:
         print(">>BEGIN: relaxedScan3D_omp()")
     if verbose > 0:
@@ -117,7 +117,7 @@ def relaxedScan3D_omp(xTips, yTips, zTips, trj=None, bF3d=False, spline_paramete
     rTips[:, :, :, 0] = xTips[:, None, None]
     rTips[:, :, :, 1] = yTips[None, :, None]
     rTips[:, :, :, 2] = zTips[::-1][None, None, :]
-    core.relaxTipStrokes_omp(rTips, rs, fs, spline_parameters=spline_parameters)
+    core.relaxTipStrokes_omp(rTips, rs, fs, tip_spline=tip_spline)
     # rs[:,:,:,:] = rs[:,:,::-1,:].transpose(2,1,0,3).copy()
     # fs[:,:,:,:] = fs[:,:,::-1,:]
     # fzs = fs[:,:,::-1,2].transpose(2,1,0).copy()
@@ -139,7 +139,7 @@ def perform_relaxation(
     FFboltz=None,
     FFkpfm_t0sV=None,
     FFkpfm_tVs0=None,
-    spline_parameters=None,
+    tip_spline=None,
     bPPdisp=False,
     bFFtotDebug=False,
     parameters=None,
@@ -176,7 +176,7 @@ def perform_relaxation(
     if parameters.tiltedScan:
         trj = trjByDir(len(zTips), d=parameters.scanTilt, p0=[0.0, 0.0, parameters.scanMin[2] - lvec[0, 2]])
     # fzs, PPpos = relaxedScan3D(xTips - lvec[0, 0], yTips - lvec[0, 1], zTips - lvec[0, 2], trj=trj, bF3d=parameters.tiltedScan)
-    fzs, PPpos = relaxedScan3D_omp(xTips - lvec[0, 0], yTips - lvec[0, 1], zTips - lvec[0, 2], trj=trj, bF3d=parameters.tiltedScan, spline_parameters=spline_parameters)
+    fzs, PPpos = relaxedScan3D_omp(xTips - lvec[0, 0], yTips - lvec[0, 1], zTips - lvec[0, 2], trj=trj, bF3d=parameters.tiltedScan, tip_spline=tip_spline)
 
     # transform probe-particle positions back to the original coordinates
     PPpos[:, :, :, 0] += lvec[0, 0]

--- a/ppafm/cli/relaxed_scan.py
+++ b/ppafm/cli/relaxed_scan.py
@@ -2,6 +2,7 @@
 
 import itertools as it
 import os
+import sys
 
 import numpy as np
 
@@ -139,6 +140,18 @@ def main(argv=None):
     lvec[2, :] = rotate_vector(lvec[2, :], opt_dict["rotate"])
     common.lvec2params(parameters=parameters, lvec=lvec)
 
+    if args.tipspline is not None:
+        try:
+            spline_parameters = core.SplineParameters.from_file(args.tipspline)
+            print(f"Loaded tip spline from {args.tipspline}")
+            print("xs: ", spline_parameters.rff_xs)
+            print("ydys: ", spline_parameters.rff_ydys)
+        except:
+            print(f"Could not load tip spline from {args.tipspline}")
+            sys.exit(1)
+    else:
+        spline_parameters = None
+
     for charge, stiffness, voltage in it.product(charges, k_constants, voltages):
         parameters.charge = charge
         parameters.klat = stiffness
@@ -160,7 +173,7 @@ def main(argv=None):
             FFboltz=ff_boltzman,
             FFkpfm_t0sV=ff_kpfm_t0sv,
             FFkpfm_tVs0=ff_kpfm_tvs0,
-            tipspline=args.tipspline,
+            spline_parameters=spline_parameters,
             bFFtotDebug=args.bDebugFFtot,
             parameters=parameters,
         )
@@ -176,7 +189,13 @@ def main(argv=None):
             print(f" === Computing eigenvectors of dynamical matrix: which={which} ddisp={parameters.ddisp}")
             tip_positions_x, tip_positions_y, tip_positions_z, lvec_scan = common.prepareScanGrids()
             r_tips = np.array(np.meshgrid(tip_positions_x, tip_positions_y, tip_positions_z)).transpose(3, 1, 2, 0).copy()
-            evals, evecs = core.stiffnessMatrix(r_tips.reshape((-1, 3)), pp_positions.reshape((-1, 3)), which=which, ddisp=parameters.ddisp)
+            evals, evecs = core.stiffnessMatrix(
+                r_tips.reshape((-1, 3)),
+                pp_positions.reshape((-1, 3)),
+                which=which,
+                ddisp=parameters.ddisp,
+                spline_parameters=spline_parameters,
+            )
             print("vib eigenval 1 min..max : ", np.min(evals[:, 0]), np.max(evals[:, 0]))
             print("vib eigenval 2 min..max : ", np.min(evals[:, 1]), np.max(evals[:, 1]))
             print("vib eigenval 3 min..max : ", np.min(evals[:, 2]), np.max(evals[:, 2]))

--- a/ppafm/cli/relaxed_scan.py
+++ b/ppafm/cli/relaxed_scan.py
@@ -142,15 +142,15 @@ def main(argv=None):
 
     if args.tipspline is not None:
         try:
-            spline_parameters = core.SplineParameters.from_file(args.tipspline)
+            tip_spline = core.SplineParameters.from_file(args.tipspline)
             print(f"Loaded tip spline from {args.tipspline}")
-            print("xs: ", spline_parameters.rff_xs)
-            print("ydys: ", spline_parameters.rff_ydys)
+            print("xs: ", tip_spline.rff_xs)
+            print("ydys: ", tip_spline.rff_ydys)
         except:
             print(f"Could not load tip spline from {args.tipspline}")
             sys.exit(1)
     else:
-        spline_parameters = None
+        tip_spline = None
 
     for charge, stiffness, voltage in it.product(charges, k_constants, voltages):
         parameters.charge = charge
@@ -173,7 +173,7 @@ def main(argv=None):
             FFboltz=ff_boltzman,
             FFkpfm_t0sV=ff_kpfm_t0sv,
             FFkpfm_tVs0=ff_kpfm_tvs0,
-            spline_parameters=spline_parameters,
+            tip_spline=tip_spline,
             bFFtotDebug=args.bDebugFFtot,
             parameters=parameters,
         )
@@ -194,7 +194,7 @@ def main(argv=None):
                 pp_positions.reshape((-1, 3)),
                 which=which,
                 ddisp=parameters.ddisp,
-                spline_parameters=spline_parameters,
+                tip_spline=tip_spline,
             )
             print("vib eigenval 1 min..max : ", np.min(evals[:, 0]), np.max(evals[:, 0]))
             print("vib eigenval 2 min..max : ", np.min(evals[:, 1]), np.max(evals[:, 1]))

--- a/ppafm/core.py
+++ b/ppafm/core.py
@@ -150,7 +150,7 @@ def setTip(lRadial=None, kRadial=None, rPP0=None, kSpring=None, parameters=None)
     lib.setTip(lRadial, kRadial, rPP0, kSpring)
 
 
-# struct SplineParameters {
+# struct SplineParams {
 #     int      rff_n;
 #     double * rff_xs;
 #     double * rff_ydys;
@@ -164,9 +164,11 @@ class SplineParameters(Structure):
 
     def __init__(self, xs, ydys):
         super().__init__()
-        self.rff_n = len(xs)
-        self.rff_xs = np.ctypeslib.as_ctypes(xs.astype(np.double))
-        self.rff_ydys = np.ctypeslib.as_ctypes(ydys.flatten().astype(np.double))
+        self.np_rff_xs = np.ascontiguousarray(xs, dtype=np.double)
+        self.np_rff_ydys = np.ascontiguousarray(ydys, dtype=np.double)
+        self.rff_n = len(self.np_rff_xs)
+        self.rff_xs = np.ctypeslib.as_ctypes(self.np_rff_xs)
+        self.rff_ydys = np.ctypeslib.as_ctypes(self.np_rff_ydys.reshape(-1))
 
     @classmethod
     def from_file(cls, file_path):
@@ -328,9 +330,9 @@ lib.relaxTipStroke.argtypes = [c_int, c_int, c_int, array2d, array2d, array2d, P
 lib.relaxTipStroke.restype = c_int
 
 
-def relaxTipStroke(rTips, rs, fs, probeStart=1, relaxAlg=1, spline_parameters=None):
+def relaxTipStroke(rTips, rs, fs, probeStart=1, relaxAlg=1, tip_spline=None):
     n = len(rTips)
-    return lib.relaxTipStroke(probeStart, relaxAlg, n, rTips, rs, fs, spline_parameters)
+    return lib.relaxTipStroke(probeStart, relaxAlg, n, rTips, rs, fs, tip_spline)
 
 
 # int relaxTipStrokes_omp( int nx, int ny, int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_, TIP::SplineParams *splineParams )
@@ -338,9 +340,9 @@ lib.relaxTipStrokes_omp.argtypes = [c_int, c_int, c_int, c_int, c_int, array4d, 
 lib.relaxTipStrokes_omp.restype = c_int
 
 
-def relaxTipStrokes_omp(rTips, rs, fs, probeStart=1, relaxAlg=1, spline_parameters=None):
+def relaxTipStrokes_omp(rTips, rs, fs, probeStart=1, relaxAlg=1, tip_spline=None):
     nx, ny, nz, _ = rTips.shape
-    return lib.relaxTipStrokes_omp(nx, ny, probeStart, relaxAlg, nz, rTips, rs, fs, spline_parameters)
+    return lib.relaxTipStrokes_omp(nx, ny, probeStart, relaxAlg, nz, rTips, rs, fs, tip_spline)
 
 
 # void stiffnessMatrix( double ddisp, int which, int n, double * rTips_, double * rPPs_, double * eigenvals_, double * evec1_, double * evec2_, double * evec3_, TIP::SplineParams *sp )
@@ -348,7 +350,7 @@ lib.stiffnessMatrix.argtypes = [c_double, c_int, c_int, array2d, array2d, array2
 lib.stiffnessMatrix.restype = None
 
 
-def stiffnessMatrix(rTips, rPPs, which=0, ddisp=0.05, spline_parameters=None):
+def stiffnessMatrix(rTips, rPPs, which=0, ddisp=0.05, tip_spline=None):
     print("py.core.stiffnessMatrix() ")
     n = len(rTips)
     eigenvals = np.zeros((n, 3))
@@ -360,7 +362,7 @@ def stiffnessMatrix(rTips, rPPs, which=0, ddisp=0.05, spline_parameters=None):
         evecs[i] = np.zeros((n, 3))
     print("py.core.stiffnessMatrix() 1 ")
 
-    lib.stiffnessMatrix(ddisp, which, n, rTips, rPPs, eigenvals, evecs[0], evecs[1], evecs[2], spline_parameters)
+    lib.stiffnessMatrix(ddisp, which, n, rTips, rPPs, eigenvals, evecs[0], evecs[1], evecs[2], tip_spline)
     return eigenvals, evecs
 
 

--- a/ppafm/core.py
+++ b/ppafm/core.py
@@ -1,6 +1,6 @@
 #!/usr/bin/python
 
-from ctypes import c_double, c_int
+from ctypes import POINTER, Structure, c_double, c_int
 
 import numpy as np
 
@@ -150,14 +150,28 @@ def setTip(lRadial=None, kRadial=None, rPP0=None, kSpring=None, parameters=None)
     lib.setTip(lRadial, kRadial, rPP0, kSpring)
 
 
-# void setTipSpline( int n, double * xs, double * ydys ){
-lib.setTipSpline.argtypes = [c_int, array1d, array2d]
-lib.setTipSpline.restype = None
+# struct SplineParameters {
+#     int      rff_n;
+#     double * rff_xs;
+#     double * rff_ydys;
+# }
+class SplineParameters(Structure):
+    _fields_ = [
+        ("rff_n", c_int),
+        ("rff_xs", POINTER(c_double)),
+        ("rff_ydys", POINTER(c_double)),
+    ]
 
+    def __init__(self, xs, ydys):
+        super().__init__()
+        self.rff_n = len(xs)
+        self.rff_xs = np.ctypeslib.as_ctypes(xs.astype(np.double))
+        self.rff_ydys = np.ctypeslib.as_ctypes(ydys.flatten().astype(np.double))
 
-def setTipSpline(xs, ydys):
-    n = len(xs)
-    lib.setTipSpline(n, xs, ydys)
+    @classmethod
+    def from_file(cls, file_path):
+        spline_data = np.genfromtxt(file_path)
+        return cls(spline_data[:, 0], spline_data[:, 1:])
 
 
 # void getInPoints_LJ( int npoints, double * points, double * FEs, int natoms_, double * Ratoms_, double * cLJs ){
@@ -309,32 +323,32 @@ def getDensityR4spline(Rs, cRAs, bNormalize=True):
     lib.getDensityR4spline(natom, Rs, cRAs)
 
 
-# int relaxTipStroke ( int probeStart, int nstep, double * rTips_, double * rs_, double * fs_ )
-lib.relaxTipStroke.argtypes = [c_int, c_int, c_int, array2d, array2d, array2d]
+# int relaxTipStroke( int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_, TIP::SplineParams *splineParams )
+lib.relaxTipStroke.argtypes = [c_int, c_int, c_int, array2d, array2d, array2d, POINTER(SplineParameters)]
 lib.relaxTipStroke.restype = c_int
 
 
-def relaxTipStroke(rTips, rs, fs, probeStart=1, relaxAlg=1):
+def relaxTipStroke(rTips, rs, fs, probeStart=1, relaxAlg=1, spline_parameters=None):
     n = len(rTips)
-    return lib.relaxTipStroke(probeStart, relaxAlg, n, rTips, rs, fs)
+    return lib.relaxTipStroke(probeStart, relaxAlg, n, rTips, rs, fs, spline_parameters)
 
 
-# int relaxTipStrokes ( int nx, int ny, int probeStart, int nstep, double * rTips_, double * rs_, double * fs_ )
-lib.relaxTipStrokes_omp.argtypes = [c_int, c_int, c_int, c_int, c_int, array4d, array4d, array4d]
+# int relaxTipStrokes_omp( int nx, int ny, int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_, TIP::SplineParams *splineParams )
+lib.relaxTipStrokes_omp.argtypes = [c_int, c_int, c_int, c_int, c_int, array4d, array4d, array4d, POINTER(SplineParameters)]
 lib.relaxTipStrokes_omp.restype = c_int
 
 
-def relaxTipStrokes_omp(rTips, rs, fs, probeStart=1, relaxAlg=1):
+def relaxTipStrokes_omp(rTips, rs, fs, probeStart=1, relaxAlg=1, spline_parameters=None):
     nx, ny, nz, _ = rTips.shape
-    return lib.relaxTipStrokes_omp(nx, ny, probeStart, relaxAlg, nz, rTips, rs, fs)
+    return lib.relaxTipStrokes_omp(nx, ny, probeStart, relaxAlg, nz, rTips, rs, fs, spline_parameters)
 
 
-# void stiffnessMatrix( double ddisp, int which, int n,  double * rTips_, double * rs_,    double * eigenvals_, double * evec1_, double * evec2_, double * evec3_ ){
-lib.stiffnessMatrix.argtypes = [c_double, c_int, c_int, array2d, array2d, array2d, array2d, array2d, array2d]
+# void stiffnessMatrix( double ddisp, int which, int n, double * rTips_, double * rPPs_, double * eigenvals_, double * evec1_, double * evec2_, double * evec3_, TIP::SplineParams *sp )
+lib.stiffnessMatrix.argtypes = [c_double, c_int, c_int, array2d, array2d, array2d, array2d, array2d, array2d, POINTER(SplineParameters)]
 lib.stiffnessMatrix.restype = None
 
 
-def stiffnessMatrix(rTips, rPPs, which=0, ddisp=0.05):
+def stiffnessMatrix(rTips, rPPs, which=0, ddisp=0.05, spline_parameters=None):
     print("py.core.stiffnessMatrix() ")
     n = len(rTips)
     eigenvals = np.zeros((n, 3))
@@ -346,7 +360,7 @@ def stiffnessMatrix(rTips, rPPs, which=0, ddisp=0.05):
         evecs[i] = np.zeros((n, 3))
     print("py.core.stiffnessMatrix() 1 ")
 
-    lib.stiffnessMatrix(ddisp, which, n, rTips, rPPs, eigenvals, evecs[0], evecs[1], evecs[2])
+    lib.stiffnessMatrix(ddisp, which, n, rTips, rPPs, eigenvals, evecs[0], evecs[1], evecs[2], spline_parameters)
     return eigenvals, evecs
 
 
@@ -378,14 +392,14 @@ def subsample_nonuniform_spline(xs, ydys, xs_, ys_=None):
     return ys_
 
 
-# void test_force( int type, int n, double * r0_, double * dr_, double * R_, double * fs_ ){
-lib.test_force.argtypes = [c_int, c_int, array1d, array1d, array1d, array2d]
+# void test_force( int type, int n, double * r0_, double * dr_, double * R_, double * fs_, TIP::SplineParams *splineParams )
+lib.test_force.argtypes = [c_int, c_int, array1d, array1d, array1d, array2d, POINTER(SplineParameters)]
 lib.test_force.restype = None
 
 
-def test_force(typ, r0, dr, R, fs):
+def test_force(typ, r0, dr, R, fs, spline_parameters=None):
     n = len(fs)
-    lib.test_force(typ, n, r0, dr, R, fs)
+    lib.test_force(typ, n, r0, dr, R, fs, spline_parameters)
     return fs
 
 

--- a/ppafm/cpp/ProbeParticle.cpp
+++ b/ppafm/cpp/ProbeParticle.cpp
@@ -54,10 +54,12 @@ namespace TIP{
     double  lRadial;    // radial PP-tip distance
     double  kRadial;    // radial PP-tip stiffness
 
-    // tip forcefiled spline
-    int      rff_n    = 0;
-    double * rff_xs   = NULL;
-    double * rff_ydys = NULL;
+    // Tip force field spline
+    struct SplineParams {
+        int      rff_n;
+        double * rff_xs;
+        double * rff_ydys;
+    };
 
     void makeConsistent(){ // place rPP0 on the sphere to be consistent with radial spring
         if( fabs(kRadial) > 1e-8 ){
@@ -65,6 +67,7 @@ namespace TIP{
             printf(" rPP0 %f %f %f \n", rPP0.x, rPP0.y, rPP0.z );
         }
     }
+
 }
 
 // relaxation namespace
@@ -206,15 +209,16 @@ inline double addAtom_Coulomb_dz2( Vec3d dR, Vec3d& fout, double * coefs ){
 }
 
 // radial spring constrain
-Vec3d forceRSpline( const Vec3d& dR, int n, double * xs, double * ydys ){
+Vec3d forceRSpline( const Vec3d& dR, TIP::SplineParams *params ){
     double x     =  sqrt( dR.norm2() );
-    int    i     = Spline_Hermite::find_index<double>( i, n-i, x, xs );
-    double x0    = xs[i  ];
-    double x1    = xs[i+1];
+    int    i     = Spline_Hermite::find_index<double>( 0, params->rff_n, x, params->rff_xs );
+    double x0    = params->rff_xs[i  ];
+    double x1    = params->rff_xs[i+1];
     double dx    = x1-x0;
     double denom = 1/dx;
     double u     = (x-x0)*denom;
     i=i<<1;
+    double *ydys = params->rff_ydys;
     double fr =  Spline_Hermite::val<double>( u, ydys[i], ydys[i+2], ydys[i+1]*dx, ydys[i+3]*dx );
     Vec3d f; f.set_mul( dR, fr/x );
     return f;
@@ -275,28 +279,28 @@ inline void evalCell( int ibuff, const Vec3d& rProbe, void * args ){
 
 // ========== Interpolations
 
-inline void getPPforce( const Vec3d& rTip, const Vec3d& r, Vec3d& f ){
+inline void getPPforce( const Vec3d& rTip, const Vec3d& r, Vec3d& f, TIP::SplineParams *splineParams ){
     Vec3d rGrid,drTip;
     rGrid.set( r.dot( gridShape.diCell.a ), r.dot( gridShape.diCell.b ), r.dot( gridShape.diCell.c ) );     // transform position from cartesian world coordinates to coordinates along which Force-Field data are sampled ( non-orthogonal cell )
     drTip.set_sub( r, rTip );                                                             // vector between Probe-particle and tip apex
     f.set    ( interpolate3DvecWrap( gridF, gridShape.n, rGrid ) );                          // force from surface, interpolated from Force-Field data array
-    if( TIP::rff_xs ){
-        f.add( forceRSpline( drTip, TIP::rff_n, TIP::rff_xs, TIP::rff_ydys ) );			  // force from tip - radial component spline
+    if( splineParams ){
+        f.add( forceRSpline( drTip, splineParams ) );                   // force from tip - radial component spline
     }else{
-        f.add( forceRSpring( drTip, TIP::kRadial, TIP::lRadial ) );                       // force from tip - radial component harmonic
+        f.add( forceRSpring( drTip, TIP::kRadial, TIP::lRadial ) );     // force from tip - radial component harmonic
     }
     drTip.sub( TIP::rPP0 );
     f.add_mul( drTip, TIP::kSpring );      // spring force
 }
 
 // relax probe particle position "r" given on particular position of tip (rTip) and initial position "r"
-int relaxProbe( int relaxAlg, const Vec3d& rTip, Vec3d& r ){
+int relaxProbe( int relaxAlg, const Vec3d& rTip, Vec3d& r, TIP::SplineParams *splineParams ){
     Vec3d v; v.set( 0.0 );
     int iter;
     FIREstate fire;
     fire.setup();
     for( iter=0; iter<RELAX::maxIters; iter++ ){
-        Vec3d f;  getPPforce( rTip, r, f );
+        Vec3d f;  getPPforce( rTip, r, f, splineParams);
         if( relaxAlg == 1 ){                                                                  // move by either damped-leap-frog ( 0 ) or by FIRE ( 1 )
             //FIRE::move( f, r, v );
             fire.move( f, r, v );
@@ -360,13 +364,6 @@ DLLEXPORT void setTip( double lRad, double kRad, double * rPP0, double * kSpring
     TIP::rPP0.set(rPP0);
     TIP::kSpring.set(kSpring);
     TIP::makeConsistent();  // rPP0 to be consistent with  lRadial
-}
-
-// set parameters of the tip like stiffness and equlibirum position in radial and lateral direction
-DLLEXPORT void setTipSpline( int n, double * xs, double * ydys ){
-    TIP::rff_n    = n;
-    TIP::rff_xs   = xs;
-    TIP::rff_ydys = ydys;
 }
 
 DLLEXPORT void getInPoints_LJ( int npoints, double * points_, double * FEs, int natoms, double * Ratoms_, double * cLJs ){
@@ -623,7 +620,7 @@ DLLEXPORT void getDensityR4spline( int natoms_, double * Ratoms_, double * cRAs 
 // returns position of probe-particle after relaxation in 1D array "rs_" and force between surface probe particle in this relaxed position in 1D array "fs_"
 // for efficiency, starting position of ProbeParticle in new point (next postion of Tip) is derived from relaxed postion of ProbeParticle from previous point
 // there are several strategies how to do it which are choosen by parameter probeStart
-DLLEXPORT int relaxTipStroke( int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_ ){
+DLLEXPORT int relaxTipStroke( int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_, TIP::SplineParams *splineParams ){
     Vec3d * rTips = (Vec3d*) rTips_;
     Vec3d * rs    = (Vec3d*) rs_;
     Vec3d * fs    = (Vec3d*) fs_;
@@ -647,7 +644,7 @@ DLLEXPORT int relaxTipStroke( int probeStart, int relaxAlg, int nstep, double * 
             rProbe.set_add( rTip, drp    );
         }
         // relax Probe Particle postion
-        int itr = relaxProbe( relaxAlg, rTip, rProbe );
+        int itr = relaxProbe( relaxAlg, rTip, rProbe, splineParams );
         if( itr>RELAX::maxIters ){
             printf( " not converged in %i iterations \n", RELAX::maxIters );
             printf( "exiting \n" );	break;
@@ -671,14 +668,14 @@ DLLEXPORT int relaxTipStroke( int probeStart, int relaxAlg, int nstep, double * 
 // returns position of probe-particle after relaxation in 1D array "rs_" and force between surface probe particle in this relaxed position in 1D array "fs_"
 // for efficiency, starting position of ProbeParticle in new point (next postion of Tip) is derived from relaxed postion of ProbeParticle from previous point
 // there are several strategies how to do it which are choosen by parameter probeStart
-DLLEXPORT int relaxTipStrokes_omp( int nx, int ny, int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_ ){
+DLLEXPORT int relaxTipStrokes_omp( int nx, int ny, int probeStart, int relaxAlg, int nstep, double * rTips_, double * rs_, double * fs_, TIP::SplineParams *splineParams ){
     printf( "relaxTipStrokes_omp()  nx %i ny %i nstep %i \n", nx, ny, nstep );
     int ndone=0;
     #pragma omp parallel for collapse(2) shared( nx, ny, probeStart, relaxAlg, nstep, rTips_, rs_, fs_, ndone )
     for (int ix=0; ix<nx; ix++){
         for (int iy=0; iy<ny; iy++){
             int ioff = (ix + iy*nx)*nstep;
-            relaxTipStroke( probeStart, relaxAlg, nstep, rTips_+ioff*3, rs_+ioff*3, fs_+ioff*3 );
+            relaxTipStroke( probeStart, relaxAlg, nstep, rTips_+ioff*3, rs_+ioff*3, fs_+ioff*3, splineParams );
             if( omp_get_thread_num()==0 ){
                 ndone++;
                 if( ndone%100==0 ){
@@ -692,7 +689,7 @@ DLLEXPORT int relaxTipStrokes_omp( int nx, int ny, int probeStart, int relaxAlg,
     return 0;
 }
 
-DLLEXPORT void stiffnessMatrix( double ddisp, int which, int n, double * rTips_, double * rPPs_, double * eigenvals_, double * evec1_, double * evec2_, double * evec3_ ){
+DLLEXPORT void stiffnessMatrix( double ddisp, int which, int n, double * rTips_, double * rPPs_, double * eigenvals_, double * evec1_, double * evec2_, double * evec3_, TIP::SplineParams *sp ){
     //printf( "C++ stiffnessMatrix() n=%i \n", n );
     Vec3d * rTips     = (Vec3d*) rTips_;
     Vec3d * rPPs      = (Vec3d*) rPPs_;
@@ -712,9 +709,9 @@ DLLEXPORT void stiffnessMatrix( double ddisp, int which, int n, double * rTips_,
         Mat3d dynmat;
         //pmin.setIfLower(rPP); pmax.setIfGreater(rPP);
         // eval dynamical matrix    D_xy = df_y/dx    = ( f(r0+dx).y - f(r0-dx).y ) / (2*dx)
-        rPP.x-=ddisp; getPPforce( rTip, rPP, f1 ); rPP.x+=2*ddisp; getPPforce( rTip, rPP, f2 );  rPP.x-=ddisp; dynmat.a.set_sub(f2,f1); dynmat.a.mul(-0.5/ddisp);
-        rPP.y-=ddisp; getPPforce( rTip, rPP, f1 ); rPP.y+=2*ddisp; getPPforce( rTip, rPP, f2 );  rPP.y-=ddisp; dynmat.b.set_sub(f2,f1); dynmat.b.mul(-0.5/ddisp);
-        rPP.z-=ddisp; getPPforce( rTip, rPP, f1 ); rPP.z+=2*ddisp; getPPforce( rTip, rPP, f2 );  rPP.z-=ddisp; dynmat.c.set_sub(f2,f1); dynmat.c.mul(-0.5/ddisp);
+        rPP.x-=ddisp; getPPforce( rTip, rPP, f1, sp ); rPP.x+=2*ddisp; getPPforce( rTip, rPP, f2, sp );  rPP.x-=ddisp; dynmat.a.set_sub(f2,f1); dynmat.a.mul(-0.5/ddisp);
+        rPP.y-=ddisp; getPPforce( rTip, rPP, f1, sp ); rPP.y+=2*ddisp; getPPforce( rTip, rPP, f2, sp );  rPP.y-=ddisp; dynmat.b.set_sub(f2,f1); dynmat.b.mul(-0.5/ddisp);
+        rPP.z-=ddisp; getPPforce( rTip, rPP, f1, sp ); rPP.z+=2*ddisp; getPPforce( rTip, rPP, f2, sp );  rPP.z-=ddisp; dynmat.c.set_sub(f2,f1); dynmat.c.mul(-0.5/ddisp);
         // symmetrize - to make sure that our symmetric matrix solver work properly
         double tmp;
         tmp = 0.5*(dynmat.xy + dynmat.yx); dynmat.xy = tmp; dynmat.yx = tmp;
@@ -775,7 +772,7 @@ DLLEXPORT void subsample_nonuniform_spline( int n, double * xs, double * ydys, i
     }
 }
 
-DLLEXPORT void test_force( int type, int n, double * r0_, double * dr_, double * R_, double * fs_ ){
+DLLEXPORT void test_force( int type, int n, double * r0_, double * dr_, double * R_, double * fs_, TIP::SplineParams *splineParams ){
     Vec3d r,dr,R;
     r .set( r0_[0], r0_[1], r0_[2] );
     dr.set( dr_[0], dr_[1], dr_[2] );
@@ -785,8 +782,8 @@ DLLEXPORT void test_force( int type, int n, double * r0_, double * dr_, double *
         //Vec3d drTip.set_sub( r, R );
         Vec3d f;
         switch( type ){
-            case 1 : f = forceRSpline( r-R, TIP::rff_n,   TIP::rff_xs, TIP::rff_ydys ); break;
-            case 2 : f = forceRSpring( r-R, TIP::kRadial, TIP::lRadial               ); break;
+            case 1 : f = forceRSpline( r-R, splineParams               ); break;
+            case 2 : f = forceRSpring( r-R, TIP::kRadial, TIP::lRadial ); break;
         }
         fs[i] = f;
         //printf( " %i (%3.3f,%3.3f,%3.3f) (%3.3f,%3.3f,%3.3f) \n", i, r.x, r.y, r.z,  f.x, f.y, f.z );

--- a/tests/human_eye/test_TipForce.py
+++ b/tests/human_eye/test_TipForce.py
@@ -7,46 +7,53 @@ import numpy as np
 
 from ppafm import common, core
 
-spline_ini = """
-0.0   0.4 -0.1
-2.0   0.2 -0.1
-4.0   0.0 -0.1
-5.0  -4.0 -4.0
-10.0 -24.0 -4.0
-"""
-# S = np.genfromtxt('TipRSpline.ini')
-S = np.genfromtxt(StringIO(spline_ini), skip_header=1)
 
-parameters = common.PpafmParameters()
+def test_tip_force():
 
-print("TipRSpline.ini overrides harmonic tip")
-xs = S[:, 0].copy()
-print("xs: ", xs)
-ydys = S[:, 1:].copy()
-print("ydys: ", ydys)
-plt.plot(xs, ydys[:, 0], "o")
-plt.plot(xs, ydys[:, 0] + ydys[:, 1], ".")
-core.setTipSpline(xs, ydys)
+    spline_ini = """
+    0.0   0.4 -0.1
+    2.0   0.2 -0.1
+    4.0   0.0 -0.1
+    5.0  -4.0 -4.0
+    10.0 -24.0 -4.0
+    """
+    # S = np.genfromtxt('TipRSpline.ini')
+    S = np.genfromtxt(StringIO(spline_ini), skip_header=1)
 
-core.setTip(parameters=parameters)
+    parameters = common.PpafmParameters()
 
-fs = np.zeros((60, 3))
-r0 = np.array([0.0, 0.0, 0.5])
-dr = np.array([0.0, 0.0, 0.1])
-R = np.array([0.0, 0.0, 0.0])
-xs = np.array(list(range(len(fs)))) * dr[2] + r0[2]
-# print "xs=",xs
+    print("TipRSpline.ini overrides harmonic tip")
+    xs = S[:, 0].copy()
+    print("xs: ", xs)
+    ydys = S[:, 1:].copy()
+    print("ydys: ", ydys)
+    plt.plot(xs, ydys[:, 0], "o")
+    plt.plot(xs, ydys[:, 0] + ydys[:, 1], ".")
+    core.setTipSpline(xs, ydys)
 
-print(">>>  core.test_force( 1, r0, dr, R, fs )")
-core.test_force(1, r0, dr, R, fs)
-plt.plot(xs, fs[:, 2])
+    core.setTip(parameters=parameters)
 
-print(">>>  core.test_force( 2, r0, dr, R, fs )")
-core.test_force(2, r0, dr, R, fs)
-plt.plot(xs, fs[:, 2])
+    fs = np.zeros((60, 3))
+    r0 = np.array([0.0, 0.0, 0.5])
+    dr = np.array([0.0, 0.0, 0.1])
+    R = np.array([0.0, 0.0, 0.0])
+    xs = np.array(list(range(len(fs)))) * dr[2] + r0[2]
+    # print "xs=",xs
 
-# print "fs:", fs
+    print(">>>  core.test_force( 1, r0, dr, R, fs )")
+    core.test_force(1, r0, dr, R, fs)
+    plt.plot(xs, fs[:, 2])
 
-plt.grid()
-plt.savefig("tip_force.png")
-plt.show()
+    print(">>>  core.test_force( 2, r0, dr, R, fs )")
+    core.test_force(2, r0, dr, R, fs)
+    plt.plot(xs, fs[:, 2])
+
+    # print "fs:", fs
+
+    plt.grid()
+    plt.savefig("tip_force.png")
+    plt.show()
+
+
+if __name__ == "__main__":
+    test_tip_force()

--- a/tests/human_eye/test_TipForce.py
+++ b/tests/human_eye/test_TipForce.py
@@ -29,7 +29,7 @@ def test_tip_force():
     print("ydys: ", ydys)
     plt.plot(xs, ydys[:, 0], "o")
     plt.plot(xs, ydys[:, 0] + ydys[:, 1], ".")
-    core.setTipSpline(xs, ydys)
+    spline_parameters = core.SplineParameters(xs, ydys)
 
     core.setTip(parameters=parameters)
 
@@ -41,11 +41,11 @@ def test_tip_force():
     # print "xs=",xs
 
     print(">>>  core.test_force( 1, r0, dr, R, fs )")
-    core.test_force(1, r0, dr, R, fs)
+    core.test_force(1, r0, dr, R, fs, spline_parameters)
     plt.plot(xs, fs[:, 2])
 
     print(">>>  core.test_force( 2, r0, dr, R, fs )")
-    core.test_force(2, r0, dr, R, fs)
+    core.test_force(2, r0, dr, R, fs, spline_parameters)
     plt.plot(xs, fs[:, 2])
 
     # print "fs:", fs

--- a/tests/human_eye/test_TipForce.py
+++ b/tests/human_eye/test_TipForce.py
@@ -17,21 +17,15 @@ def test_tip_force():
     5.0  -4.0 -4.0
     10.0 -24.0 -4.0
     """
-    # S = np.genfromtxt('TipRSpline.ini')
-    S = np.genfromtxt(StringIO(spline_ini), skip_header=1)
 
-    parameters = common.PpafmParameters()
+    tip_spline = core.SplineParameters.from_file(StringIO(spline_ini))
 
-    print("TipRSpline.ini overrides harmonic tip")
-    xs = S[:, 0].copy()
-    print("xs: ", xs)
-    ydys = S[:, 1:].copy()
-    print("ydys: ", ydys)
-    plt.plot(xs, ydys[:, 0], "o")
-    plt.plot(xs, ydys[:, 0] + ydys[:, 1], ".")
-    spline_parameters = core.SplineParameters(xs, ydys)
+    print("xs: ", tip_spline.np_rff_xs)
+    print("ydys: ", tip_spline.np_rff_ydys)
+    plt.plot(tip_spline.np_rff_xs, tip_spline.np_rff_ydys[:, 0], "o")
+    plt.plot(tip_spline.np_rff_xs, tip_spline.np_rff_ydys[:, 0] + tip_spline.np_rff_ydys[:, 1], ".")
 
-    core.setTip(parameters=parameters)
+    core.setTip(parameters=common.PpafmParameters())
 
     fs = np.zeros((60, 3))
     r0 = np.array([0.0, 0.0, 0.5])
@@ -41,11 +35,11 @@ def test_tip_force():
     # print "xs=",xs
 
     print(">>>  core.test_force( 1, r0, dr, R, fs )")
-    core.test_force(1, r0, dr, R, fs, spline_parameters)
+    core.test_force(1, r0, dr, R, fs, tip_spline)
     plt.plot(xs, fs[:, 2])
 
     print(">>>  core.test_force( 2, r0, dr, R, fs )")
-    core.test_force(2, r0, dr, R, fs, spline_parameters)
+    core.test_force(2, r0, dr, R, fs, tip_spline)
     plt.plot(xs, fs[:, 2])
 
     # print "fs:", fs

--- a/tests/human_eye/test_splines.py
+++ b/tests/human_eye/test_splines.py
@@ -6,46 +6,53 @@ import numpy as np
 
 import ppafm.core as PPC
 
-# =========== uniform spline
 
-ys = np.array([0.0, 0.2, -0.3, 0.0])
-dys = np.array([0.8, 0.8, -1.2, -1.2])
-xs_ = np.linspace(-0.89, 0.59, 100)
-ydys = np.transpose(np.array([ys, dys])).copy()
-# print "ydys" = ydys
+def test_splines():
 
-x0 = -0.9
-dx = 0.5
-ys_ = PPC.subsample_uniform_spline(x0, dx, ydys, xs_)
-xs = np.array(list(range(len(ys)))) * dx + x0
+    # =========== uniform spline
 
-plt.figure()
-plt.plot(xs, ys, "o")
-plt.plot(xs, ys + 0.1 * dys, ".")
-plt.plot(xs_, ys_, "-")
-plt.grid()
+    ys = np.array([0.0, 0.2, -0.3, 0.0])
+    dys = np.array([0.8, 0.8, -1.2, -1.2])
+    xs_ = np.linspace(-0.89, 0.59, 100)
+    ydys = np.transpose(np.array([ys, dys])).copy()
+    # print "ydys" = ydys
 
-plt.savefig("splines1.png")
+    x0 = -0.9
+    dx = 0.5
+    ys_ = PPC.subsample_uniform_spline(x0, dx, ydys, xs_)
+    xs = np.array(list(range(len(ys)))) * dx + x0
 
-# =========== non-uniform spline
+    plt.figure()
+    plt.plot(xs, ys, "o")
+    plt.plot(xs, ys + 0.1 * dys, ".")
+    plt.plot(xs_, ys_, "-")
+    plt.grid()
 
-xs = np.array([-0.9, -0.6, 0.4, 0.6])
-# ys  = np.array   ( [  0.0, 0.2, -0.3,  0.0 ] )
-# dys = np.array  ( [  0.8, 0.8, -1.2, -1.2 ] )
-# dys = np.array   ( [  0.0, 0.0, 0.0, 0.0   ] )
-# xs_ = np.linspace(   -0.89, 0.59, 20           )
+    plt.savefig("splines1.png")
 
-ys_ = PPC.subsample_nonuniform_spline(xs, ydys, xs_)
+    # =========== non-uniform spline
 
-plt.figure()
-plt.plot(xs, ys, "o")
-plt.plot(xs, ys + 0.1 * dys, ".")
-plt.plot(xs_, ys_, "-")
-plt.grid()
+    xs = np.array([-0.9, -0.6, 0.4, 0.6])
+    # ys  = np.array   ( [  0.0, 0.2, -0.3,  0.0 ] )
+    # dys = np.array  ( [  0.8, 0.8, -1.2, -1.2 ] )
+    # dys = np.array   ( [  0.0, 0.0, 0.0, 0.0   ] )
+    # xs_ = np.linspace(   -0.89, 0.59, 20           )
 
-# print "xs_", xs_
-# print "ys_", ys_
+    ys_ = PPC.subsample_nonuniform_spline(xs, ydys, xs_)
 
-plt.savefig("splines2.png")
+    plt.figure()
+    plt.plot(xs, ys, "o")
+    plt.plot(xs, ys + 0.1 * dys, ".")
+    plt.plot(xs_, ys_, "-")
+    plt.grid()
 
-plt.show()
+    # print "xs_", xs_
+    # print "ys_", ys_
+
+    plt.savefig("splines2.png")
+
+    plt.show()
+
+
+if __name__ == "__main__":
+    test_splines()

--- a/tests/human_eye/test_vdwDamping.py
+++ b/tests/human_eye/test_vdwDamping.py
@@ -8,115 +8,116 @@ import numpy as np
 import ppafm.common as PPU
 import ppafm.core as PPC
 
-# ======== Setup
-iPP = 8
-iZs = [1, 6, 7, 8]
-# iZs   = [1]
-iAtom = 3
 
-iVdWModel = 3
-if len(sys.argv) > 1:
-    try:
-        iVdWModel = int(sys.argv[1])
-    except ValueError:
-        print(f"Invalid input argument {sys.argv[1]}")
+def test_vdw_damping():
 
-ADamps = [180.0, 0.5, 0.5, 0.03, 0.01]
+    # ======== Setup
+    iPP = 8
+    iZs = [1, 6, 7, 8]
+    # iZs   = [1]
+    iAtom = 3
 
-model_names = [
-    #   "LJ_RE",     # -3
-    #   "LJ   ",     # -2
-    #   "invR6",     # -1
-    "VdW_C6",  # 0
-    "VdW_R2",  # 1
-    "VdW_R4",  # 2
-    "VdW_invR4",  # 3
-    "VdW_invR8",  # 4
-]
+    iVdWModel = 3
+    if len(sys.argv) > 1:
+        try:
+            iVdWModel = int(sys.argv[1])
+        except ValueError:
+            print(f"Invalid input argument {sys.argv[1]}")
+
+    ADamps = [180.0, 0.5, 0.5, 0.03, 0.01]
+
+    model_names = [
+        #   "LJ_RE",     # -3
+        #   "LJ   ",     # -2
+        #   "invR6",     # -1
+        "VdW_C6",  # 0
+        "VdW_R2",  # 1
+        "VdW_R4",  # 2
+        "VdW_invR4",  # 3
+        "VdW_invR8",  # 4
+    ]
+
+    # ======== Functions
+
+    def plotDecor(x0, vmax):
+        plt.axvline(x0, ls=":", c="k")
+        plt.axhline(0, ls="-", c="k")
+        plt.ylim(vmax * -2, vmax)
+        plt.grid()
+
+    def deriv(xs, Es, d=None):
+        if d is None:
+            d = xs[1] - xs[0]  # ;print( "d ", d )
+        Fs = -(Es[2:] - Es[:-2]) / (2.0 * d)
+        xs = xs[1:-1]
+        return Fs, xs
+
+    # ======== Main
+
+    FFparams = PPU.loadSpecies()
+    PPU.getFFdict(FFparams)
+    # print elem_dict
+
+    REs = PPU.getAtomsRE(iPP, iZs, FFparams)
+    print("REs  ", REs)
+    cLJs = PPU.getAtomsLJ(iPP, iZs, FFparams)
+    print("cLJs ", cLJs)
+
+    xs = np.arange(0.0, 6.0, 0.1)  # ;print(xs)
+
+    """
+    Es_LJ,Fs_LJ = PPC.evalRadialFF( xs, REs[iAtom,:], kind=-3 ) #;print( "Es_LJ", Es_LJ )  # LJ  -C6/r^6 + C12/r^12
+    #Es_LJ,Fs_LJ = PPC.evalRadialFF( xs, cLJs[iAtom,:], kind=-2 ) ;print( "Es_LJ", Es_LJ )  # LJ  -C6/r^6 + C12/r^12
+    Es_r6,Fs_r6 = PPC.evalRadialFF( xs, cLJs[iAtom,:], kind=0 ) #;print( "Es_r6", Es_r6 )   # vdW -C6/r^6
+
+    Es,Fs       = PPC.evalRadialFF( xs, cLJs[iAtom,:], kind=0  )
+    #Es,Fs       = PPC.evalRadialFF( xs, REs [iAtom,:], kind=0  )
+
+    plt.figure(figsize=(8,16))
+    plt.subplot(2,1,1);    plt.plot(xs,Es_LJ   ,':k'); plt.plot(xs,Es_r6   ,'--k'); plt.plot(xs,Es   );       plotDecor( R0, E0*2 )
+    plt.subplot(2,1,2);    plt.plot(xs,Fs_LJ*-1,':k'); plt.plot(xs,Fs_r6*-1,'--k'); plt.plot(xs,Fs*-1);       plotDecor( R0, E0*5 )
+    """
+
+    coefs = REs
+    if iVdWModel in {-2, -1, 0}:
+        coefs = cLJs
+
+    name = model_names[iVdWModel]
+    nz = len(iZs)
+    plt.figure(figsize=(5 * nz, 5 * 2))
+    for i in range(nz):
+        iAtom = i
+
+        R0 = REs[iAtom, 0]
+        E0 = REs[iAtom, 1]
+        Es_LJ, Fs_LJ = PPC.evalRadialFF(-xs, REs[iAtom, :], kind=-3)  # ;print( "Es_LJ", Es_LJ )  # LJ  -C6/r^6 + C12/r^12
+        Es_r6, Fs_r6 = PPC.evalRadialFF(-xs, cLJs[iAtom, :], kind=-1)  # ;print( "Es_r6", Es_r6 )   # vdW -C6/r^6
+        ADamp = ADamps[iVdWModel]
+        Es, Fs = PPC.evalRadialFF(-xs, coefs[iAtom, :], kind=iVdWModel, ADamp=ADamp)
+
+        Fs_num, xs_ = deriv(xs, Es)
+
+        plt.subplot(2, nz, i + 1)
+        plt.plot(xs, Es_LJ, ":k")
+        plt.plot(xs, Es_r6, "--k")
+        plt.plot(xs, Es)
+        plotDecor(R0, E0 * 5)
+        plt.title(f"{name}(Adamp={ADamp:5.2f}) {(FFparams[iZs[i] - 1][4]).decode()}-{(FFparams[iPP - 1][4]).decode()}")  # ; plt.title("Interaction %i-%i" %(iZs[i], iPP) )
+        plt.subplot(2, nz, i + nz + 1)
+        plt.plot(xs, Fs_LJ, ":k")
+        plt.plot(xs, Fs_r6, "--k")
+        plt.plot(xs, Fs)
+        plt.plot(xs_, Fs_num, ":r")
+        plotDecor(R0, E0 * 7)
+
+    plt.subplot(2, nz, +1)
+    plt.ylabel("Energy [eV]")
+    plt.subplot(2, nz, nz + 1)
+    plt.ylabel("Force  [eV/A]")
+    plt.savefig(name + ".png", bbox_inches="tight")
+
+    plt.show()
 
 
-# ======== Functions
-
-
-def plotDecor(x0, vmax):
-    plt.axvline(x0, ls=":", c="k")
-    plt.axhline(0, ls="-", c="k")
-    plt.ylim(vmax * -2, vmax)
-    plt.grid()
-
-
-def deriv(xs, Es, d=None):
-    if d is None:
-        d = xs[1] - xs[0]  # ;print( "d ", d )
-    Fs = -(Es[2:] - Es[:-2]) / (2.0 * d)
-    xs = xs[1:-1]
-    return Fs, xs
-
-
-# ======== Main
-
-FFparams = PPU.loadSpecies()
-elem_dict = PPU.getFFdict(FFparams)
-# print elem_dict
-
-REs = PPU.getAtomsRE(iPP, iZs, FFparams)
-print("REs  ", REs)
-cLJs = PPU.getAtomsLJ(iPP, iZs, FFparams)
-print("cLJs ", cLJs)
-
-
-xs = np.arange(0.0, 6.0, 0.1)  # ;print(xs)
-
-"""
-Es_LJ,Fs_LJ = PPC.evalRadialFF( xs, REs[iAtom,:], kind=-3 ) #;print( "Es_LJ", Es_LJ )  # LJ  -C6/r^6 + C12/r^12
-#Es_LJ,Fs_LJ = PPC.evalRadialFF( xs, cLJs[iAtom,:], kind=-2 ) ;print( "Es_LJ", Es_LJ )  # LJ  -C6/r^6 + C12/r^12
-Es_r6,Fs_r6 = PPC.evalRadialFF( xs, cLJs[iAtom,:], kind=0 ) #;print( "Es_r6", Es_r6 )   # vdW -C6/r^6
-
-Es,Fs       = PPC.evalRadialFF( xs, cLJs[iAtom,:], kind=0  )
-#Es,Fs       = PPC.evalRadialFF( xs, REs [iAtom,:], kind=0  )
-
-plt.figure(figsize=(8,16))
-plt.subplot(2,1,1);    plt.plot(xs,Es_LJ   ,':k'); plt.plot(xs,Es_r6   ,'--k'); plt.plot(xs,Es   );       plotDecor( R0, E0*2 )
-plt.subplot(2,1,2);    plt.plot(xs,Fs_LJ*-1,':k'); plt.plot(xs,Fs_r6*-1,'--k'); plt.plot(xs,Fs*-1);       plotDecor( R0, E0*5 )
-"""
-
-coefs = REs
-if iVdWModel in {-2, -1, 0}:
-    coefs = cLJs
-
-
-name = model_names[iVdWModel]
-nz = len(iZs)
-plt.figure(figsize=(5 * nz, 5 * 2))
-for i in range(nz):
-    iAtom = i
-
-    R0 = REs[iAtom, 0]
-    E0 = REs[iAtom, 1]
-    Es_LJ, Fs_LJ = PPC.evalRadialFF(-xs, REs[iAtom, :], kind=-3)  # ;print( "Es_LJ", Es_LJ )  # LJ  -C6/r^6 + C12/r^12
-    Es_r6, Fs_r6 = PPC.evalRadialFF(-xs, cLJs[iAtom, :], kind=-1)  # ;print( "Es_r6", Es_r6 )   # vdW -C6/r^6
-    ADamp = ADamps[iVdWModel]
-    Es, Fs = PPC.evalRadialFF(-xs, coefs[iAtom, :], kind=iVdWModel, ADamp=ADamp)
-
-    Fs_num, xs_ = deriv(xs, Es)
-
-    plt.subplot(2, nz, i + 1)
-    plt.plot(xs, Es_LJ, ":k")
-    plt.plot(xs, Es_r6, "--k")
-    plt.plot(xs, Es)
-    plotDecor(R0, E0 * 5)
-    plt.title(f"{name}(Adamp={ADamp:5.2f}) {(FFparams[iZs[i] - 1][4]).decode()}-{(FFparams[iPP - 1][4]).decode()}")  # ; plt.title("Interaction %i-%i" %(iZs[i], iPP) )
-    plt.subplot(2, nz, i + nz + 1)
-    plt.plot(xs, Fs_LJ, ":k")
-    plt.plot(xs, Fs_r6, "--k")
-    plt.plot(xs, Fs)
-    plt.plot(xs_, Fs_num, ":r")
-    plotDecor(R0, E0 * 7)
-
-plt.subplot(2, nz, +1)
-plt.ylabel("Energy [eV]")
-plt.subplot(2, nz, nz + 1)
-plt.ylabel("Force  [eV/A]")
-plt.savefig(name + ".png", bbox_inches="tight")
-
-plt.show()
+if __name__ == "__main__":
+    test_vdw_damping()

